### PR TITLE
Allow octets type to accept string and integer values for parsing

### PR DESCRIPTION
**Pull request by nward**
The juniper/ERX dictionary from the FreeRADIUS project includes using octets attributes with values, which are written in the dictionary as integers, i.e.:
```
ATTRIBUTE	ERX-Acct-Request-Reason			210	octets
VALUE	ERX-Acct-Request-Reason		Acct-Start-Ack		1
VALUE	ERX-Acct-Request-Reason		Periodic		2
VALUE	ERX-Acct-Request-Reason		IP-Active		4
...etc...
```

pyrad doesn't parse this at present, and errors out as `"1"` in python is not of bytes type.

This PR adds this functionality, by allowing either str or bytes starting with `0x`, and checking for digits-only values, and packing them as an integer to get a bytes output the same as if the `0x` integer value had been passed.

Additionally, in the `EncodeOctets` function (and many Encode* function in tools.py) the parameter is called `str`, which conflicts with the python basic string type. In the EncodeOctets function I have renamed this to `octetstring`.

### CODE REVIEW

**Comment by nward** [comment]
I've subsequently added a fix for #152, and tests for it. Thanks!

File: pyrad/tools.py [hunk 1]
```diff
@@ -18,19 +18,26 @@ def EncodeString(str):
 
 
 def EncodeOctets(octetstring):
-    if len(octetstring) > 253:
+    # Check for max length of the hex encoded with 0x prefix, as a sanity check
+    if len(octetstring) > 508:
         raise ValueError('Can only encode strings of <= 253 characters')
 
     if type(octetstring) == bytes and octetstring.startswith(b'0x'):
```
Comment (nerosketch) [inline pyrad/tools.py:25]: Python has 'isinstance' for it. Like this:
isinstance(octetstring, bytes)

File: pyrad/tools.py [hunk 2]
```diff
@@ -18,19 +18,26 @@ def EncodeString(str):
...
         hexstring = octetstring.split(b'0x')[1]
-        return binascii.unhexlify(hexstring)
```
Comment (nerosketch) [inline pyrad/tools.py]: isinstance(octetstring, str)

File: pyrad/tools.py [hunk 3]
```diff
@@ -18,19 +18,26 @@ def EncodeString(str):
...
+        encoded_octets = binascii.unhexlify(hexstring)
     elif type(octetstring) == str and octetstring.startswith('0x'):
         hexstring = octetstring.split('0x')[1]
-        return binascii.unhexlify(hexstring)
+        encoded_octets = binascii.unhexlify(hexstring)
     elif type(octetstring) == str and octetstring.isdecimal():
```
Comment (nerosketch) [inline pyrad/tools.py:31]: isinstance(octetstring, str)

**Review by nerosketch — CHANGES_REQUESTED** [review]

File: pyrad/tools.py [hunk 4]
```diff
@@ -17,15 +17,20 @@ def EncodeString(str):
...
 
-def EncodeOctets(str):
-    if len(str) > 253:
+def EncodeOctets(octetstring):
+    if len(octetstring) > 253:
         raise ValueError('Can only encode strings of <= 253 characters')
 
-    if str.startswith(b'0x'):
-        hexstring = str.split(b'0x')[1]
+    if type(octetstring) == bytes and octetstring.startswith(b'0x'):
```
Comment (nerosketch) [inline pyrad/tools.py:25]: Python has 'isinstance' for it. Like this:
`if isinstance(octetstring, bytes) and octetstring.startswith(b'0x'):`

File: pyrad/tools.py [hunk 5]
```diff
@@ -17,15 +17,20 @@ def EncodeString(str):
...
+        hexstring = octetstring.split(b'0x')[1]
+        return binascii.unhexlify(hexstring)
+    elif type(octetstring) == str and octetstring.startswith('0x'):
```
Comment (nerosketch) [inline pyrad/tools.py:28]: `isinstance(octetstring, str)`

File: pyrad/tools.py [hunk 6]
```diff
@@ -17,15 +17,20 @@ def EncodeString(str):
...
+        hexstring = octetstring.split('0x')[1]
         return binascii.unhexlify(hexstring)
+    elif type(octetstring) == str and octetstring.isdecimal():
```
Comment (nerosketch) [inline pyrad/tools.py:31]: `isinstance(octetstring, str)`

**Review by nerosketch — CHANGES_REQUESTED** [review]

**Comment by nerosketch** [comment]
I need these changes too. I also have "juniper", and I have similar error.

**Comment by nward** [comment]
Thanks @nerosketch - I have no idea why I used type() rather than isinstance.. brain fart perhaps!

**Comment by Istvan91** [comment]
looks good to me - I'll try to push a new version this weekend.

**Pull request merged**